### ISSUE TITLE
clip-path-polygon transitions

### ISSUE THREAD

**Issue by sebastiangrochocki** [andrusieczko/clip-path-polygon#14]
Hello! Is there any possibility to make some transitions for clip-path-polygon?
I have tried something like that but it does not work:

jQuery(function($) {
var points = [[350, 0], [0, 350], [350, 350]];
$('#mask').clipPath(points);

$(".item-112").mouseover(function() {
        points = [[0, 0], [0, 350], [350, 0]];
        $('#mask').css('-moz-transition','all 2s ease 0s').clipPath(points);
})

$(".item-112").mouseout(function() {
        var points = [[350, 0], [0, 350], [350, 350]];
        $('#mask').css('-moz-transition','all 2s ease 0s').clipPath(points);
})
});

Regards

**Comment by andrusieczko** [comment]
Hi Sebastian,

Sorry for a late response but time around Christmas and New Year is quite busy :)

Thanks to you I discovered one bug related to passing options, I'll fix it soon.

Regarding your issue, I made it working with Chrome, unfortunately I couldn't have forced it to cooperate with Firefox yet.

Will keep you updated.

Thanks again!
Karol

**Comment by gtsop** [comment]
I'm going to use this lib on my project and the transition support is a must.  I was wondering weather you've got any updates on this feature, else I could submit a pull request provided that I find a solution.

**Comment by andrusieczko** [comment]
Hi @gtsop,

I haven't progressed with this issue any further.
Your contribution would be highly appreciated! :)

Notice: I will be unreachable for next couple of days.

Thanks!
Karol

**Comment by gtsop** [comment]
@andrusieczko 

I ended up using a brute force method (lazy me), and it works with a barely acceptable performance penalty. I pre-calculate all intermediate clip points, cache the DOM Element, cache the options object and use an interval to call .clipPath() within the required transition time.

However, while trying to find a better solution I came across [this](http://stackoverflow.com/a/25372255) which is obviously better. Haven't tried implementing it yet, just sharing it here to give a lead

**Comment by andrusieczko** [comment]
Hi @gtsop,

Thanks for sharing!

I'll try to take a look at that this weekend.

Cheers,
Karol

**Comment by manniL** [comment]
Is there something new regarding this issue? @andrusieczko

**Comment by andrusieczko** [comment]
Hi all,

There are a couple of things to consider.

# Chrome

@sebastiangrochocki's solution works well in *Chrome*:

JS:
```javascript
let pointsFrom = [[350, 0], [0, 350], [350, 350]];
let pointsTo = [[0, 0], [0, 350], [350, 0]];
$('#myClipPathDiv').clipPath(pointsFrom);

$(".animatebutton").mouseover(function() {
  $('#myClipPathDiv').clipPath(points, pointsTo);
});
$(".animatebutton").mouseout(function() {
  $('#myClipPathDiv').clipPath(points, pointsFrom);
});
```
HTML:
```html
<div style="width: 350px; height: 350px; padding: 10px">
  <div id="myClipPathDiv"></div>
</div>
<span id="animatebutton">Animate!</span>
```
CSS:
```css
#myClipPathDiv {
  transition: clip-path 1s;
  width: 100%;
  height: 100%;
  background-color: green;
}
```

# Firefox

Unfortunately, the situation with *Firefox* is a bit different...
*Firefox* doesn't support `transform` in indirectly rendered things like `clippath` (source: https://bugzilla.mozilla.org/show_bug.cgi?id=1118710#c13).

However, they support *Synchronized Multimedia Integration Language* (*SMIL*).
Unfortunately, *Chrome 45* deprecated *SMIL* in favor of *CSS animations* and *Web animations* (source: https://developer.mozilla.org/en-US/docs/Web/SVG/SVG_animation_with_SMIL).

This is how you can achieve the desired result with *SMIL* in Firefox:

JS:
```javascript
$('#mask').clipPath([[350, 0], [0, 350], [350, 350]], {
  svgDefId: 'myClipPathSvg'
});
```
HTML:
```html
<div style="width: 350px; height: 350px; padding: 10px">
  <div id="myClipPathDiv"></div>
</div>
<span id="animatebutton">Animate!</span>

<svg width="0" height="0" style="position: absolute; visibility: hidden; width: 0px; height: 0px;">
  <defs>
    <clipPath id="myClipPathSvg">
      <polygon points="350 0, 0 350, 350 350">
        <animate attributeName="points" fill="freeze" from="350 0, 0 350, 350 350" to="0 0, 0 350, 350 0" dur="1s" begin="animatebutton.mouseover"></animate>
        <animate attributeName="points" fill="freeze" from="0 0, 0 350, 350 0" to="350 0, 0 350, 350 350" dur="1s" begin="animatebutton.mouseout"></animate>
      </polygon>
    </clipPath>
  </defs>
</svg>
```
CSS:
```css
#myClipPathDiv {
  width: 100%;
  height: 100%;
  background-color: green;
}
```

# TL;DR

feature         | Chrome | Firefox
------------ | -------- | --------
*transform* | supported | not supported
*SMIL*         | not supported | supported

# Solution

Having said all of that, you'd probably need to implement two different solutions for *Chrome* and *Firefox* (or, to be more precise, WebKit/Blink and Gecko).

I will not add this feature to the library itself as it would increase the complexity of it.

I'm sorry I hadn't resolved it earlier.

We might look at [Web Animations](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API) in the future.

@sebastiangrochocki, @gtsop, @manniL: does it answer your questions?

**Comment by manniL** [comment]
Indeed, it does! Thanks for taking the time and especially thank you for the detailed answer

**Comment by gtsop** [comment]
@andrusieczko I'm good with that answer, thanks 👍
